# Diabetes Prediction System

## Meta Description
A machine learning model that analyzes patient medical data to predict whether a **female patient** has diabetes, using all 8 diagnostic features from the Pima Indians Diabetes Database with MLP Neural Network and StandardScaler preprocessing.

## Introduction

**Definition:** A binary classification ML task that predicts whether a female patient aged 21 or older is Diabetic (1) or Not Diabetic (0) based on 8 medical input features.

**Application:** Used in clinical decision support, early diabetes screening, and healthcare data analytics specifically for female patients of Pima Indian heritage aged 21 and above.

**Dataset Note:** This model is trained on the Pima Indians Diabetes Database, originally from the National Institute of Diabetes and Digestive and Kidney Diseases. All patients in this dataset are **females at least 21 years old** of Pima Indian heritage. Therefore, predictions are only valid for this demographic.

## Input and Output

### Inputs (Features / Attributes)

| Input Feature | Description | Example Values |
|---|---|---|
| Pregnancies | Number of times pregnant | 0, 1, 2, 6 ... |
| Glucose | Plasma glucose concentration (mg/dL) | 85, 120, 148 ... |
| BloodPressure | Diastolic blood pressure (mm Hg) | 60, 72, 80 ... |
| SkinThickness | Triceps skin fold thickness (mm) | 0, 20, 35 ... |
| Insulin | 2-Hour serum insulin (mu U/ml) | 0, 80, 200 ... |
| BMI | Body Mass Index (kg/m²) | 22.5, 33.6, 43.1 ... |
| DiabetesPedigreeFunction | Diabetes pedigree function score | 0.167, 0.627, 1.2 ... |
| Age | Age in years (minimum 21) | 21, 35, 50 ... |

### Output (Prediction)

| Output Attribute | Possible Values |
|---|---|
| Outcome | 1 (Diabetic), 0 (Not Diabetic) |

### Examples (Sample Dataset)

| Pregnancies | Glucose | BloodPressure | SkinThickness | Insulin | BMI | DiabetesPedigreeFunction | Age | Outcome |
|:-----------:|:-------:|:-------------:|:-------------:|:-------:|:---:|:------------------------:|:---:|:-------:|
| 6 | 148 | 72 | 35 | 0 | 33.6 | 0.627 | 50 | 1 |
| 1 | 85 | 66 | 29 | 0 | 26.6 | 0.351 | 31 | 0 |
| 0 | 137 | 40 | 35 | 168 | 43.1 | 2.288 | 33 | 1 |

## Developer & System Information

| | |
|---|---|
| **Dataset** | Pima Indians Diabetes Database (UCI / Kaggle) |
| **Target Population** | Females aged 21 and above |
| **Total Columns** | 9 (8 input features + 1 output) |
| **Sample Size** | 100 instances (50 Diabetic, 50 Not Diabetic) |
| **IDE** | Jupyter Notebook |
| **Programming Language** | Python 3 |
| **Libraries** | NumPy, Pandas, Pickle, scikit-learn, PrettyTable |
| **Algorithm** | MLP Neural Network + StandardScaler |

## Table of Contents
- Step 1: Import Libraries
- Step 2: Load Sample Data
- Step 3: Understand and Pre-process Sample Data
  - Step 3.1: Understand Sample Data
  - Step 3.2: Pre-process Sample Data (StandardScaler)
- Step 4: Feature Extraction
- Step 5: Label Encoding
- Step 6: Execute the Training Phase
  - Step 6.1: Split into Training and Testing Data
  - Step 6.2: Split Input Vectors and Labels of Training Data
  - Step 6.3: Scale the Features using StandardScaler
  - Step 6.4: Train the MLP Neural Network Classifier
  - Step 6.5: Save the Trained Model and Scaler
- Step 7: Execute the Testing Phase
  - Step 7.1: Split Input Vectors and Labels of Testing Data
  - Step 7.2: Load the Saved Model and Scaler
  - Step 7.3: Make Predictions on Testing Data
  - Step 7.4: Calculate the Accuracy Score
- Step 8: Execute the Application Phase
  - Step 8.1: Take Input from User
  - Step 8.2: Convert User Input into Feature Vector
  - Step 8.3: Scale the Feature Vector
  - Step 8.4: Load the Saved Model
  - Step 8.5: Model Prediction
- Step 9: Execute the Feedback Phase

## Step 1: Import Libraries

In [ ]:
"""
Purpose of this code:
----------------------
This script imports the required libraries for building and evaluating
a machine learning model for diabetes prediction. It sets up tools for:
- Data manipulation and loading
- Splitting data into training and testing sets
- Scaling features using StandardScaler
- Training an MLP Neural Network
- Evaluating prediction accuracy
- Displaying results in tabular form
"""

# Import numerical computing library
import numpy as np

# Import data manipulation library
import pandas as pd

# Import pickle for saving and loading trained models
import pickle

# Import train_test_split to divide dataset into training and testing sets
from sklearn.model_selection import train_test_split

# Import StandardScaler to normalize feature values (required for MLP)
from sklearn.preprocessing import StandardScaler

# Import MLP Neural Network classifier
from sklearn.neural_network import MLPClassifier

# Import accuracy_score to evaluate model performance
from sklearn.metrics import accuracy_score

# Import PrettyTable for clean tabular output
from prettytable import PrettyTable

print("All libraries imported successfully!")

All libraries imported successfully!


## Step 2: Load Sample Data

In [ ]:
"""
Purpose of this section:
-------------------------
Load the 100-instance sample dataset (50 Diabetic, 50 Not Diabetic) from
a CSV file into a pandas DataFrame and display it for review.
The dataset contains all 8 input features and 1 output column (Outcome).
"""

# Read the CSV file into a pandas DataFrame
sample_data = pd.read_csv("sample-data.csv")

# Configure pandas to display all rows and columns without truncation
pd.set_option("display.max_rows", None, "display.max_columns", None)

print("\n\nSample Data:")
print("============\n")
print(sample_data)



Sample Data:

    Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0             1       97             70             40        0  38.1   
1             5       78             48              0        0  33.7   
2             3      111             58             31       44  29.5   
3             2      129             84              0        0  28.0   
4             7      102             74             40      105  37.2   
5             1      124             74             36        0  27.8   
6             1      119             44             47       63  35.5   
7             1      105             58              0        0  24.3   
8             1       96            122              0        0  22.4   
9             3      130             64              0        0  23.1   
10            1       97             66             15      140  23.2   
11            3       99             80             11       64  19.3   
12            6      183           

## Step 3: Understand and Pre-process Sample Data
### Step 3.1: Understand Sample Data

In [ ]:
"""
Purpose of this section:
-------------------------
Understand the dataset by printing:
1. All attribute (column) names
2. Total number of instances
3. Class distribution (Diabetic vs Not Diabetic)
4. Data types and basic statistics
"""

print("\n\nAttributes in Sample Data:")
print("==========================\n")
print(sample_data.columns.tolist())

print("\n\nNumber of Instances in Sample Data:", sample_data["Pregnancies"].count())
print("=============================================\n")

print("\nOutcome Distribution (0 = Not Diabetic, 1 = Diabetic):")
print("=========================================================\n")
print(sample_data["Outcome"].value_counts())

print("\n\nData Types of Each Column:")
print("===========================\n")
print(sample_data.dtypes)

print("\n\nBasic Descriptive Statistics:")
print("==============================\n")
print(sample_data.describe())



Attributes in Sample Data:

['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome']


Number of Instances in Sample Data: 100


Outcome Distribution (0 = Not Diabetic, 1 = Diabetic):

Outcome
0    50
1    50
Name: count, dtype: int64


Data Types of Each Column:

Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object


Basic Descriptive Statistics:

       Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   100.000000  100.000000     100.000000     100.000000  100.000000   
mean      4.030000  122.320000      70.050000      22.120000   78.900000   
std       3.313441   32.348444      18.057358      16.205922  107.

### Step 3.2: Pre-process Sample Data — StandardScaler

**Why StandardScaler is needed:**
- All 8 features have very different numeric ranges:
  - Insulin ranges from 0 to 846
  - DiabetesPedigreeFunction ranges from 0.08 to 2.42
  - Pregnancies ranges from 0 to 17
- MLP Neural Networks are sensitive to feature scale — features with large ranges dominate and cause poor predictions.
- StandardScaler transforms each feature to have **mean = 0** and **standard deviation = 1**, putting all features on equal footing.

Scaling is applied in **Step 6.3** after the train/test split (to prevent data leakage).

## Step 4: Feature Extraction
- All **8 features** from the dataset are used as inputs — no feature extraction needed.


## Step 5: Label Encoding
- All 8 input features are already **numeric** — no Label Encoding required for inputs.
- The output column (Outcome) is already **0 or 1** — no Label Encoding required for output.


## Step 6: Execute the Training Phase
### Step 6.1: Splitting Sample Data into Training Data and Testing Data

In [ ]:
"""
Purpose of this section:
-------------------------
Split the 100-instance dataset into:
- Training Data: 80 instances (80%) — used to TEACH the model
- Testing Data:  20 instances (20%) — used to CHECK model performance

random_state=0 : ensures reproducibility (same split every run)
shuffle=False  : keeps original row order before splitting
"""

# Load directly from sample data (no encoding needed)

# 80% training, 20% testing
training_data, testing_data = train_test_split(
    sample_data,
    test_size=0.2,
    random_state=0,
    shuffle=True
)

# Save both splits to CSV
training_data.to_csv("training-data.csv", index=False, header=True)
testing_data.to_csv("testing-data.csv",   index=False, header=True)

print("\n\nTraining Data (80 instances):")
print("==============================\n")
print(training_data)

print("\n\nTesting Data (20 instances):")
print("=============================\n")
print(testing_data)



Training Data (80 instances):

    Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
43            8      112             72              0        0  23.6   
62            9      112             82             32      175  34.2   
3             2      129             84              0        0  28.0   
71            7      100              0              0        0  30.0   
45            2      107             74             30      100  33.6   
48            2      111             60              0        0  26.2   
6             1      119             44             47       63  35.5   
99            6      134             70             23      130  35.4   
82           10      168             74              0        0  38.0   
76            7      114             64              0        0  27.4   
60           13      152             90             33       29  26.8   
80            0      113             76              0        0  33.3   
90            9   

### Step 6.2: Splitting Input Vectors and Outputs / Labels of Training Data

In [ ]:
"""
Purpose of this section:
-------------------------
Separate the training data into:
- Input Vectors (X_train): All 8 feature columns
- Output Labels (y_train): Outcome column (1=Diabetic, 0=Not Diabetic)
"""

print("\n\nInput Vectors (Feature Vectors) of Training Data:")
print("==================================================\n")
input_vector_train = training_data.iloc[:, :-1]
print(input_vector_train)

print("\n\nOutputs / Labels of Training Data:")
print("====================================\n")
print("  Outcome")
output_label_train = training_data.iloc[:, -1]
print(output_label_train)



Input Vectors (Feature Vectors) of Training Data:

    Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
43            8      112             72              0        0  23.6   
62            9      112             82             32      175  34.2   
3             2      129             84              0        0  28.0   
71            7      100              0              0        0  30.0   
45            2      107             74             30      100  33.6   
48            2      111             60              0        0  26.2   
6             1      119             44             47       63  35.5   
99            6      134             70             23      130  35.4   
82           10      168             74              0        0  38.0   
76            7      114             64              0        0  27.4   
60           13      152             90             33       29  26.8   
80            0      113             76              0        0  33.3  

### Step 6.3: Scale the Features using StandardScaler

In [ ]:
"""
Purpose of this section:
-------------------------
Apply StandardScaler to normalize all 8 input features.

Why scale AFTER splitting?
- We fit the scaler ONLY on training data (fit_transform).
- We then apply the same scaler to test data (transform only).
- This prevents data leakage — the model never sees test data statistics
  during training, which would give artificially inflated accuracy.

fit_transform(X_train) : learns mean/std from training data AND scales it
transform(X_test)      : applies the SAME learned scaling to test data
"""

# Initialize the StandardScaler
scaler = StandardScaler()

# Fit on training data and scale it
input_vector_train_scaled = scaler.fit_transform(input_vector_train)

# Scale testing data using the SAME scaler (do NOT fit again)
input_vector_test  = testing_data.iloc[:, :-1]
input_vector_test_scaled = scaler.transform(input_vector_test)

print("\n\nScaled Input Vectors of Training Data:")
print("=======================================\n")
print(input_vector_train_scaled)

print("\n\nScaled Input Vectors of Testing Data:")
print("======================================\n")
print(input_vector_test_scaled)



Scaled Input Vectors of Training Data:

[[ 1.19931224e+00 -2.38000952e-01  1.21542716e-01 -1.40890904e+00
  -8.35940535e-01 -1.20240351e+00  1.29790032e+00  2.58376991e+00]
 [ 1.50293559e+00 -2.38000952e-01  7.10840730e-01  5.63101425e-01
   1.23505000e+00  2.23464448e-01 -7.72884869e-01  3.90120590e-01]
 [-6.22427872e-01  2.95246650e-01  8.28700333e-01 -1.40890904e+00
  -8.35940535e-01 -6.10533793e-01 -6.87197206e-01 -5.07281406e-01]
 [ 8.95688888e-01 -6.14411024e-01 -4.12140299e+00 -1.40890904e+00
  -8.35940535e-01 -3.41502103e-01  2.68666527e-02 -8.72474163e-03]
 [-6.22427872e-01 -3.94838482e-01  2.39402318e-01  4.39850771e-01
   3.47482626e-01  1.42754941e-01 -2.58758891e-01 -9.06126738e-01]
 [-6.22427872e-01 -2.69368458e-01 -5.85614902e-01 -1.40890904e+00
  -8.35940535e-01 -8.52662315e-01 -4.76548368e-01 -9.06126738e-01]
 [-9.26051224e-01 -1.84284098e-02 -1.52849173e+00  1.48748133e+00
  -9.03839439e-02  3.98335047e-01 -7.01478483e-01 -7.06704072e-01]
 [ 5.92065536e-01  4.520841

### Step 6.4: Train the MLP Neural Network Classifier

In [ ]:
"""
Purpose of this section:
-------------------------
Train an MLP Neural Network on the scaled training data.

MLP learns patterns through interconnected layers of neurons. It outperforms SVC on this dataset
because it can capture non-linear relationships between medical features.

Parameters used:
- max_iter=1000   : allows enough training iterations to converge
- alpha=1         : L2 regularization to prevent overfitting on small dataset
- random_state=0  : ensures reproducible results
"""

print("\n\nTraining the MLP Neural Network Classifier on Training Data")
print("=============================================================\n")

# Initialize MLP Neural Network
mlp_model = MLPClassifier(max_iter=1000, alpha=1, random_state=0)

# Train the model on SCALED training data
mlp_model.fit(input_vector_train_scaled, np.ravel(output_label_train))

print("Model trained successfully!")
print("\nModel Parameters:")
print("=================\n")
print(mlp_model)



Training the MLP Neural Network Classifier on Training Data

Model trained successfully!

Model Parameters:

MLPClassifier(alpha=1, max_iter=1000, random_state=0)


### Step 6.5: Save the Trained Model and Scaler

In [ ]:
"""
Purpose of this section:
-------------------------
Save both the trained MLP model AND the fitted StandardScaler to disk.

IMPORTANT: Both must be saved together because:
- The scaler learned the mean/std from training data.
- Any new input MUST be scaled using this SAME scaler before prediction.
- If we lose the scaler, the model cannot make correct predictions.
"""

# Save the trained MLP model
pickle.dump(mlp_model, open("mlp_trained_model.pkl", "wb"))

# Save the fitted scaler (must use same scaler for new predictions)
pickle.dump(scaler, open("scaler.pkl", "wb"))

print("\nMLP model saved as: mlp_trained_model.pkl")
print("Scaler saved as:    scaler.pkl")


MLP model saved as: mlp_trained_model.pkl
Scaler saved as:    scaler.pkl


## Step 7: Execute the Testing Phase
### Step 7.1: Splitting Input Vectors and Outputs/Labels of Testing Data

In [ ]:
"""
Purpose of this section:
-------------------------
Separate the testing data into input features and output labels.
(input_vector_test_scaled was already prepared in Step 6.3)
"""

print("\n\nInput Vectors (Feature Vectors) of Testing Data - Scaled:")
print("==========================================================\n")
print(input_vector_test_scaled)

print("\n\nOutputs / Labels of Testing Data:")
print("==================================\n")
print("  Outcome")
output_label_test = testing_data.iloc[:, -1]
print(output_label_test)



Input Vectors (Feature Vectors) of Testing Data - Scaled:

[[-6.22427872e-01  7.56741083e-02  3.68311259e-03  2.54974790e-01
  -8.35940535e-01  5.73205646e-01 -4.87259326e-01 -5.07281406e-01]
 [-1.22967458e+00  2.63879144e-01 -1.14176490e-01 -2.38027825e-01
   1.29422115e+00 -2.74244180e-01  3.26514625e+00 -7.06704072e-01]
 [-3.18804520e-01 -2.69368458e-01 -7.03474505e-01  5.01476098e-01
  -3.15234344e-01 -4.08760025e-01 -1.65930589e-01 -1.00583807e+00]
 [-3.18804520e-01  1.67541692e+00  8.28700333e-01  6.24726752e-01
   4.77348525e+00  4.25238216e-01 -7.80025508e-01 -1.00583807e+00]
 [-1.51811676e-02  8.28494253e-01  1.30013874e+00 -1.40890904e+00
  -8.35940535e-01 -1.80083088e-01  2.23234214e-01  2.88290391e+00]
 [ 8.95688888e-01  2.95246650e-01 -1.14176490e-01  1.61073198e+00
   6.43338416e-01  8.01882584e-01 -1.33797716e-01  1.08809992e+00]
 [ 1.50293559e+00 -9.59453590e-01 -4.67755299e-01 -1.40890904e+00
  -8.35940535e-01 -1.35037094e+00 -1.19418255e+00  9.09865913e-02]
 [-1.229

### Step 7.2: Load the Saved Model and Scaler

In [ ]:
"""
Purpose of this section:
-------------------------
Load the saved MLP model and StandardScaler from disk.
Both are needed to make predictions on new data.
"""

model  = pickle.load(open("mlp_trained_model.pkl", "rb"))
scaler = pickle.load(open("scaler.pkl", "rb"))
print("\nMLP model loaded successfully!")
print("Scaler loaded successfully!")


MLP model loaded successfully!
Scaler loaded successfully!


### Step 7.3: Evaluate the Machine Learning Model
#### Step 7.3.1: Make Predictions with the Trained Model on Testing Data

In [ ]:
"""
Purpose of this section:
-------------------------
Use the trained MLP model to make predictions on the scaled testing data.
Store predictions alongside actual values for comparison.
"""

# Make predictions on scaled test data
model_predictions = model.predict(input_vector_test_scaled)

# Suppress chained assignment warning
pd.options.mode.chained_assignment = None

# Add predictions as a new column to testing dataset
testing_data["Predictions"] = model_predictions

# Save predictions to CSV
testing_data.to_csv("model-predictions.csv", index=False, header=True)

print("\n\nPredictions Returned by mlp_trained_model:")
print("===========================================\n")
print(testing_data)



Predictions Returned by mlp_trained_model:

    Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
26            2      122             70             27        0  36.8   
86            0      128             68             19      180  30.5   
2             3      111             58             31       44  29.5   
55            3      173             84             33      474  35.7   
75            4      146             92              0        0  31.2   
93            7      129             68             49      125  38.5   
16            9       89             62              0        0  22.5   
73            0      198             66             32      274  41.3   
54           10      161             68             23      132  25.5   
95            9      164             84             21        0  30.8   
53            7      178             84              0        0  39.9   
92            1      181             64             30      180  34.1   
78   

### Step 7.4: Calculate the Accuracy Score

In [ ]:
"""
Purpose of this section:
-------------------------
Calculate the accuracy of the MLP model on the testing dataset.
Accuracy = (Correct Predictions / Total Predictions) x 100
"""

model_accuracy_score = accuracy_score(
    testing_data["Outcome"],
    testing_data["Predictions"]
)

print("\n\nAccuracy Score:")
print("===============\n")
print(round(model_accuracy_score, 2))
print(f"\nThe MLP model correctly predicted {round(model_accuracy_score * 100, 1)}% of the 20 test instances.")



Accuracy Score:

0.8

The MLP model correctly predicted 80.0% of the 20 test instances.


## Step 8: Execute the Application Phase
### Step 8.1: Take Input from User

In [ ]:
"""
Purpose of this section:
-------------------------
Ask the user to enter their 8 medical measurements for diabetes prediction.
"""

pregnancies_input = float(input("\nPlease enter number of Pregnancies (e.g., 0, 1, 2, 6): ").strip())
glucose_input     = float(input("\nPlease enter Glucose level in mg/dL (e.g., 85, 120, 148): ").strip())
bp_input          = float(input("\nPlease enter Blood Pressure in mm Hg (e.g., 60, 72, 80): ").strip())
skin_input        = float(input("\nPlease enter Skin Thickness in mm (e.g., 0, 20, 35): ").strip())
insulin_input     = float(input("\nPlease enter Insulin level in mu U/ml (e.g., 0, 80, 200): ").strip())
bmi_input         = float(input("\nPlease enter BMI value (e.g., 22.5, 33.6, 43.1): ").strip())
dpf_input         = float(input("\nPlease enter Diabetes Pedigree Function (e.g., 0.167, 0.627, 1.2): ").strip())
age_input         = float(input("\nPlease enter Age in years (e.g., 21, 35, 50): ").strip())


Please enter number of Pregnancies (e.g., 0, 1, 2, 6): 2

Please enter Glucose level in mg/dL (e.g., 85, 120, 148): 85

Please enter Blood Pressure in mm Hg (e.g., 60, 72, 80): 80

Please enter Skin Thickness in mm (e.g., 0, 20, 35): 35

Please enter Insulin level in mu U/ml (e.g., 0, 80, 200): 150

Please enter BMI value (e.g., 22.5, 33.6, 43.1): 33.6

Please enter Diabetes Pedigree Function (e.g., 0.167, 0.627, 1.2): 0.67

Please enter Age in years (e.g., 21, 35, 50): 38


### Step 8.2: Convert User Input into Feature Vector

In [ ]:
"""
Purpose of this section:
-------------------------
Convert the user's 8 inputs into a structured DataFrame (feature vector)
that exactly matches the column format used during training.
"""

user_input = pd.DataFrame({
    "Pregnancies":              [pregnancies_input],
    "Glucose":                  [glucose_input],
    "BloodPressure":            [bp_input],
    "SkinThickness":            [skin_input],
    "Insulin":                  [insulin_input],
    "BMI":                      [bmi_input],
    "DiabetesPedigreeFunction": [dpf_input],
    "Age":                      [age_input]
})

print("\n\nUser Input Feature Vector:")
print("==========================\n")
print(user_input)



User Input Feature Vector:

   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0          2.0     85.0           80.0           35.0    150.0  33.6   

   DiabetesPedigreeFunction   Age  
0                      0.67  38.0  


### Step 8.3: Scale the Feature Vector (Exactly Same as Scaling Applied to Sample Data)

In [ ]:
"""
Purpose of this section:
-------------------------
Scale the user's input using the SAME StandardScaler that was fitted
on the training data and saved in Step 6.5.

Using the saved scaler ensures the transformation is identical to
what was applied during training — critical for correct predictions.
"""

# Load the saved scaler
scaler = pickle.load(open("scaler.pkl", "rb"))

# Scale user input using the same scaler
user_input_scaled = scaler.transform(user_input)

print("\n\nUser Input Feature Vector (Original):")
print("======================================\n")
print(user_input)

print("\n\nUser Input Feature Vector (Scaled):")
print("=====================================\n")
print(user_input_scaled)



User Input Feature Vector (Original):

   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0          2.0     85.0           80.0           35.0    150.0  33.6   

   DiabetesPedigreeFunction   Age  
0                      0.67  38.0  


User Input Feature Vector (Scaled):

[[-0.62242787 -1.08492361  0.59298113  0.74797741  0.93919421  0.14275494
   0.69094604  0.58954326]]


### Step 8.4: Load the Saved Model

In [ ]:
"""
Purpose of this section:
-------------------------
Load the saved MLP model from disk for making predictions on new user data.
"""

model = pickle.load(open("mlp_trained_model.pkl", "rb"))
print("\nMLP model loaded successfully!")


MLP model loaded successfully!


### Step 8.5: Model Prediction
#### Step 8.5.1: Apply Model on the Scaled Feature Vector and Return Prediction to the User

In [ ]:
"""
Purpose of this section:
-------------------------
Apply the trained MLP model to the scaled user input and display
whether the patient is predicted to be Diabetic or Not Diabetic.
"""

# Make prediction on scaled user input
predicted_outcome = model.predict(user_input_scaled)

# Translate numeric result to readable label
if predicted_outcome[0] == 1:
    prediction = "DIABETIC"
else:
    prediction = "NOT DIABETIC"

# Display result in PrettyTable
pretty_table = PrettyTable()
pretty_table.add_column("       ** Prediction **       ", [prediction])
print(pretty_table)

+--------------------------------+
|        ** Prediction **        |
+--------------------------------+
|          NOT DIABETIC          |
+--------------------------------+


## Step 9: Execute the Feedback Phase

### Step 9.1: Collect Feedback
After the model is deployed, feedback is collected from doctors and patients about the **accuracy, usability, and reliability** of the diabetes prediction model in real-world screening.

### Step 9.2: Possible Improvements
1. **Use a larger dataset:** Train the model on the full **768 records** instead of only 100.  
2. **Handle zero values:** Replace impossible **0 values** in features like Glucose, BMI, Insulin, BloodPressure, and SkinThickness with the **median value** of the column.  
3. **Hyperparameter tuning:** Use **GridSearchCV** to find better model parameters.  
4. **Cross-validation:** Apply **k-fold cross-validation** for more reliable accuracy.  
5. **Compare algorithms:** Test other models such as **Random Forest** and **Logistic Regression**.  
6. **Feature analysis:** Identify which of the **8 medical features** contribute most to diabetes prediction.

### Step 9.3: Improve the Model
Based on feedback, the model can be retrained using the **full dataset**, with **zero-value correction** and **GridSearchCV tuning**. These improvements may increase accuracy from **about 80% to around 85% or higher**.

---

## Conclusion
The analysis shows that diabetes prediction can be effectively performed using **8 diagnostic medical features** such as glucose level, BMI, blood pressure, and age. The **MLP Neural Network with StandardScaler preprocessing** achieved the best performance, reaching about **80% accuracy** on the sample dataset. Feature scaling was important because the variables had very different value ranges. Although the current model used only **100 records**, the results demonstrate that machine learning can assist in **early diabetes risk prediction**. Using the **full dataset, better preprocessing, and parameter tuning** can further improve the model’s accuracy and reliability.